### PR TITLE
importccl: Always run AddSSTable with disallowShadowing for IMPORT

### PR DESCRIPTION
**Pull request by adityamaru27**
This change does a few things to integrate disallowShadowing with
IMPORT:

1. It switches ON the disallowShadowing option in AddSSTable
for both direct ingest and sampling implementations of IMPORT.
Collision detection is only needed to enforce correctness of
incremental import as we will not encounter any collisions when
importing into a new table. However, the overhead of always
keeping this option set for both IMPORT and IMPORT INTO is not enough
(~1.1%) to justify plumbing a user option to toggle this flag.
If in the future this resting overhead increases, I have added a TODO
explaining the same.

2. This change slightly modifies the error handling of the C++ collision
detection logic to return human readable versions of the colliding key.

3. Following #39201, we do not consider keys with the same value and ts
as colliding in AddSSTable. This is to accomodate for IMPORT job
pause and resumption. To maintain uniform behavior, keys within the
same SST with the same value will also not be considered a duplicate
error. The subsequent duplicates will simply not be added to the SST.
It is worth noting, that this behavior is only applicable for the direct
ingest IMPORT option. The regular IMPORT will still raise duplicate key
errors. The overhead of adding it to the old IMPORT path is not
deemed worth considering we are going to do away with it soon.

### CODE REVIEW

File: pkg/ccl/importccl/sst_writer_proc.go [hunk 1]
```diff
@@ -214,7 +214,13 @@ func (sp *sstWriter) Run(ctx context.Context) {
 						// throughput.
 						log.Errorf(ctx, "failed to scatter span %s: %s", roachpb.PrettyPrintKey(nil, end), pErr)
 					}
-					if err := bulk.AddSSTable(ctx, sp.db, sst.span.Key, sst.span.EndKey, sst.data, false /* disallowShadowing */); err != nil {
+					// TODO(adityamaru): There is an observed 1.1% overhead of always
+					// disallowing shadow keys in AddSSTable i.e. when no collisions are
+					// found. Since IMPORT TABLE ingests data into a new key space there
+					// can never be any shadow keys, and this check could be turned off.
```
Comment (dt) [inline pkg/ccl/importccl/sst_writer_proc.go]: Direct-ingest import has no sort so we _could_ collide with our own keys even in new table import.

Comment (dt) [reply pkg/ccl/importccl/sst_writer_proc.go]: ugh, never mind, I got confused.

**Review by dt — APPROVED** [review]

**Comment by adityamaru27** [comment]
@dt this has become a little more involved since #39201. PTAL, thanks.

**Review by adityamaru27 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39152)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru27 and @dt)

___
*[pkg/ccl/importccl/sst_writer_proc.go, line 220 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39152#-LkzIaCMm3rxLF0ug_sq-r2-220:-Llh_i1O-L9DEyypfOpO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1527345303c2fec7260497dd37d3767204f03b73/pkg/ccl/importccl/sst_writer_proc.go#L220)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

ugh, never mind, I got confused.
</blockquote></details>

Done.
___
*[pkg/storage/bulk/sst_batcher.go, line 125 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39152#-LlDEk7ti2qOOgyUcOAi-r4-125:-Llh_tER96heTGd2QA0F:bgi9q0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/62272c01b7180949107d8f4df76b50c297bd6fbd/pkg/storage/bulk/sst_batcher.go#L125)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

😬 Is buffering every KV? 

Can we take advantage of the fact that we're looking at _sorted_ keys and just keep a copy of the value we saw on the last call?
</blockquote></details>

ugh, i don't know why I forgot the input is sorted. Done.

**Review by adityamaru27 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39152)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/storage/bulk/sst_batcher.go, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39152#-LlDEk7ti2qOOgyUcOAi-r4-49:-Llha0q-6IbIVkHWHQSh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/62272c01b7180949107d8f4df76b50c297bd6fbd/pkg/storage/bulk/sst_batcher.go#L49)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

can you add to these comments with when/where we might use these?
`e.g. when used to backfill an inverted index where <blah>"`

Partly I'm curious about when we want non-value-equal dupe skipping vs when we want value-equal skipping.
</blockquote></details>

Done.

**Review by dt — APPROVED** [review]

File: pkg/ccl/importccl/csv_testdata_helpers_test.go [hunk 2]
```diff
@@ -28,6 +29,73 @@ type csvTestFiles struct {
...
+		t.Fatal(err)
+	}
+
+	return fmt.Sprintf(`'nodelocal:///shadow-data'`)
+}
+
+func getDupWithSameValueFile(t testing.TB) string {
+	dir := filepath.Join("testdata", "csv")
+	fileName := filepath.Join(dir, "dup-key-same-value")
+	f, err := os.Create(fileName)
```
Comment (dt) [inline pkg/ccl/importccl/csv_testdata_helpers_test.go]: Did you want to checkin the file and only rewrite it based on an env var?

**Comment by adityamaru27** [comment]
TFTR!
bors r+

**Comment by adityamaru27** [comment]
I suspect some sort of flake, followed by a retry caused the testraces to fail in bors. I can't seem to see any issue in CI or when running locally.

**Pull request merged**